klaus/contrib/autodiscover.py

```diff
@@ -0,0 +1,194 @@
+                logger.info(
+                    "'%s' -> '%s'; updating references",
+                    event.src_path,
+                    event.dest_path
+                )
+                self._klausapp.add_repo(event.dest_path)
+                break
+        else:
+            logger.info(
+                "%s was moved out of KLAUS_REPOS to %s, and is now ignored",
+                event.src_path,
+                event.dest_path
+            )
+
+
+class KlausAutoDiscover(Klaus):
+    """
+    Instead of the normal Klaus behaviour of receiving a list of git
+    repositories as `os.environ['KLAUS_REPOS']`, this searches
+    `os.environ['KLAUS_REPOS']` for valid git repositories and updates klaus
+    automatically when any changes are made. if `recursive=True`, klaus will
+    watch every subdirectory in the given directories, and discover all git
+    repositories therein. This simplifies administration significantly, but may
+    cause problems with large repositories, as the underlying file watch
+    mechanisms don't expose a recursion depth parameter on any of the common
+    platforms (Linux, win32, OS X).
+    """
+    def __init__(self, basedirs, site_name, use_smarthttp, **kwargs):
+        recursive = kwargs.pop('discover_recursive', False)
+        super(KlausAutoDiscover, self).__init__([], site_name, use_smarthttp, **kwargs)
+        self._lock = threading.RLock()
+        self._basedirs = basedirs
+        self._observer = Observer()
+
+        # First find and add all the existing repos in the given base directory.
+        for directory in basedirs:
+            for fspath in os.listdir(directory):
+                path = os.path.join(directory, fspath)
+                self.add_repo(path)
+
+        # Then create a handler to notify us of any changes within those dirs.
+        fs_event_handler = KlausFSEventHandler(self)
+        for directory in basedirs:
+            self._observer.schedule(fs_event_handler, directory, recursive=recursive)
+
+        self._observer.start()
+
+    def add_repo(self, directory):
+        """
+        Attempts to register `directory` with the klaus app as a git repository.
+        If `directory` is not a git repo, this method does nothing.
+        """
+        with self._lock:
+            try:
+                repo = FancyRepo(directory)
+            except dulwich.errors.NotGitRepository:
+                logger.debug("Dulwich says '%s' is not a git repository", directory)
+                return
+
+            if self.repo_map.get(repo.name):
+                logger.debug("skipping repo with duplicate name:%s", repo)
+                return
+            logger.info("Adding repository, '%s' at '%s'", repo.name, directory)
+            self.repo_map[repo.name] = repo
+
+    def remove_repo(self, directory):
+        """
+        given a filesystem path, `directory` attempts to remove the
+        corresponding repository from the list of watched repositories.
+        """
```

Comment:
This should be split into two separate methods: One that checks whether we have the directory in our list, and one that assumes we have it in our list and removes it.